helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -835,8 +1080,28 @@ protected void processStateChanged(WatchedEvent event) {
     if (getShutdownTrigger()) {
       return;
     }
+
     fireStateChangedEvent(event.getState());
-    if (isManagingZkConnection() && event.getState() == KeeperState.Expired) {
+
+    if (!isManagingZkConnection()) {
+      return;
+    }
+
+    if (event.getState() == KeeperState.SyncConnected) {
+      if (!_isNewSessionEventFired) {
+        /*
+         * Once the new zookeeper session receives the first SyncConnected state,
+         * the zookeeper session is established and connected to zookeeper service.
+         * So the session id is available and non-zero. Now we can fire new session events.
+         */
+        fireNewSessionEvents();
+        /*
+         * Set it true to avoid firing events again for the same session next time
```

Comment:
How to ensure this flag is for the "same session" only? I didn't see you check if it is the "same session", or reset the flag.

Reply:
Resetting the flag is after reconnect:
https://github.com/apache/helix/blob/b485d1f708e3308fa46a0ce88dda36607d93e186/helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java#L1143-L1145